source/opt/type_manager.cpp

```diff
@@ -81,17 +80,17 @@ void TypeManager::RemoveId(uint32_t id) {
     for (auto& pair : id_to_type_) {
       if (pair.first != id && *pair.second == *type) {
         // Equivalent ambiguous type, re-map type.
-        type_to_id_.erase(type.get());
-        type_to_id_[pair.second.get()] = pair.first;
+        type_to_id_.erase(type);
+        type_to_id_[pair.second] = pair.first;
         found = true;
         break;
       }
       // No equivalent ambiguous type, remove mapping.
-      if (!found) type_to_id_.erase(type.get());
+      if (!found) type_to_id_.erase(type);
```

Comment:
Looks like this should be outside the for loop at line 80.  Otherwise it's going to indiscriminantly delete types. ?!

Reply:
Missing test?

Reply:
will fix